sparse/src/KokkosSparse_CcsMatrix.hpp

```diff
@@ -0,0 +1,232 @@
+/// \class CcsMatrix
+/// \brief Compressed sparse column implementation of a sparse matrix.
+/// \tparam ScalarType The type of entries in the sparse matrix.
+/// \tparam OrdinalType The type of column indices in the sparse matrix.
+/// \tparam Device The Kokkos Device type.
+/// \tparam MemoryTraits Traits describing how Kokkos manages and
+///   accesses data.  The default parameter suffices for most users.
+///
+/// "Ccs" stands for "compressed column sparse."  This is the phrase
+/// Trilinos traditionally uses to describe compressed sparse column
+/// storage for sparse matrices, as described, for example, in Saad
+/// (2nd ed.).
+template <class ScalarType, class OrdinalType, class Device,
+          class MemoryTraits = void,
+          class SizeType     = typename Kokkos::ViewTraits<OrdinalType*, Device,
+                                                       void, void>::size_type>
+class CcsMatrix {
+  static_assert(
+      std::is_signed<OrdinalType>::value,
+      "CcsMatrix requires that OrdinalType is a signed integer type.");
+
+ public:
+  //! Type of the matrix's execution space.
+  typedef typename Device::execution_space execution_space;
+  //! Type of the matrix's memory space.
+  typedef typename Device::memory_space memory_space;
+  //! Canonical device type
+  typedef Kokkos::Device<execution_space, memory_space> device_type;
+  typedef MemoryTraits memory_traits;
+
+  /// \brief Type of each entry of the "column map."
+  ///
+  /// The "column map" corresponds to the \c ptr array of column offsets in
+  /// compressed sparse column (CSR) storage.
+  typedef SizeType size_type;
+  //! Type of each value in the matrix.
+  typedef ScalarType value_type;
+  //! Type of each (column) index in the matrix.
+  typedef OrdinalType ordinal_type;
+  //! Type of the graph structure of the sparse matrix - consistent with Kokkos.
+  typedef Kokkos::StaticCrsGraph<ordinal_type, default_layout, device_type,
+                                 memory_traits, size_type>
+      staticccsgraph_type;
+  //! Type of the "column map" (which contains the offset for each column's
+  //! data).
+  typedef typename staticccsgraph_type::row_map_type col_map_type;
+  typedef Kokkos::View<value_type*, Kokkos::LayoutRight, device_type,
+                       MemoryTraits>
+      values_type;
+  //! Type of column indices in the sparse matrix.
+  typedef typename staticccsgraph_type::entries_type index_type;
+
+#ifdef KOKKOS_USE_CUSPARSE
+  cusparseHandle_t cusparse_handle;
+  cusparseMatDescr_t cusparse_descr;
+#endif  // KOKKOS_USE_CUSPARSE
+
+  /// \name Storage of the actual sparsity structure and values.
+  ///
+  /// CcsMatrix uses the compressed sparse column (CSR) storage format to
+  /// store the sparse matrix.  CSR is also called "compressed column
+  /// storage"; hence the name, which it inherits from Tpetra and from
+  /// Epetra before it.
+  //@{
+  //! The graph (sparsity structure) of the sparse matrix.
+  staticccsgraph_type graph;
```

Comment:
@e10harvey I see an issue with just renaming StaticCrsGraph to StaticCcsGraph here. I would expect to be able to do this, assuming consistency with CrsMatrix:
```
KokkosSparse::CcsMatrix<...> myMatrix(...);
doThing(myMatrix.graph.col_map, myMatrix.graph.entries);
```
but there won't be a member ``col_map`` in the graph (it'll still be called ``row_map``). To fix that, you could either provide ``Kokkos::StaticCcsGraph`` in this file, or even better, add it into Kokkos containers in a separate PR first.

Reply:
To clarify the first option here, I would be OK with temporarily adding
```
namespace Kokkos
{
template<...>
struct StaticCcsGraph
{ ... };
}
```
to this file, rather than putting this in Kokkos containers next to StaticCrsGraph.

Reply:
Thanks for your feedback, @brian-kelley! Do you have any objection to reverting https://github.com/kokkos/kokkos-kernels/pull/1446/commits/32f6281400334005710a97fec82ff24ff33d927d and then adding a `StaticCcsGraph : StaticCsGraph` and `StaticCrsGraph : StaticCsGraph` to Kokkos Core?

Reply:
Please disregard. I will add a StaticCcsGraph to KokkosKernels.